### PR TITLE
Allow dispensing objects, dispenser object customization

### LINKED ISSUE #1013: Idea: Changing the sprites of enemies that come out of dispensers.
Right now, when a dispenser dispenses an enemy, it is in its default sprite, and as far as I know, there is no way to change that. Can this be in the game so that I can change the sprites of the dispensed enemies before they come out of their dispensers? For example, the poison ivy that comes out of the dispenser could automatically have a pumpkin sprite, etc. Thanks!

### LINKED ISSUE #1421: Allow different types of things to come out of cannon, dispensers etc.
SuperTux 0.6.2
Windows 7 32-bit

Currently, you can only have enemies come out of cannons, dispensers etc. Maybe, you can also have  other things like power-ups, stars, and others stuff come out of them as well. It'll be much better. Maybe, we can also have pipes give out stuff. In this game, Pipes don't serve any purpose except for just acting as solid ground. I know that it'll require some optimizations but it'll be a good addition to the game.

### LINKED ISSUE #1706: Suggestion: Improve customisation of Dispensers
Right now, dispensers can only dispense certain badguys to their default settings. It would be cool to allow dispensers to automatically set certain properties of what they dispense. For example, the list of badguys available from the submenu could display this badguy's properties.

Since it is possible to add multiple badguys of the same type in the list (I guess it's related to the likeliness of a certain badguy to spawn), it would be easy to make it so a certain badguy can spawn with different settings.

For the `name` property exposing objects to Squirrel, it could be interesting to expose to the name not an object, but an array of the objects that will be created. Each badguy in the list could have different names or all have the same name. The list would contain all the badguys currently alive in the game, in order from oldest to youngest,

### LINKED ISSUE #2280: Idea: Dispensers can dispense objects
Dispensers should have the ability to dispense objects since dispensers can dispense enemies, so why not objects? This would be a fun feature

### PR DESCRIPTION
**Pull request by Vankata453**
`MovingObject`s (with some exceptions) can now be dispensed through dispensers! Bonus blocks, coins, coin rains, coin explosions and gold bombs, which usually add coins to the level stats, now don't, if they have been dispensed (have a parent dispenser assigned).

Any object, added to a dispenser, can now also be customized! This includes everything, other than setting a name, or manual position and width.

A flag has been added to `ObjectFactory`, which only allows certain objects to be dispensable. Others are blacklisted for reasons, including them not being `MovingObject`s, or breaking the game if dispensed.

The dispenser options for limitation `limit-dispensed-badguys` and `max-concurrent-badguys` only take effect, when a `BadGuy` is being dispensed. If a `BadGuy` is about to be dispensed, and the limit has been reached, nothing would be dispensed.

Fixes #1013.
Closes #1421.
Fixes #1706.
Closes #2280.

### CODE REVIEW

**Comment by tylerandari13** [comment]
I got a funny error.

```
Error: signal 11:
supertux2(_ZN12ErrorHandler17print_stack_traceEv+0x2f)[0x55effdb9599f]
supertux2(_ZN12ErrorHandler12handle_errorEi+0x3b)[0x55effdb95a2b]
/lib/x86_64-linux-gnu/libc.so.6(+0x42520)[0x7f2415e9f520]
supertux2(_ZN10GameObject10save_stateEv+0x21)[0x55effd98a941]
supertux2(_ZN10ObjectMenuC2EP10GameObjectRKSt8functionIFvS1_EE+0xa0)[0x55effda91e60]
supertux2(_ZN16ObjectSelectMenu11menu_actionER8MenuItem+0xc9)[0x55effdad9e49]
supertux2(_ZN4Menu14process_actionERK10MenuAction+0x231)[0x55effd935791]
supertux2(_ZN4Menu5eventERK9SDL_Event+0x141)[0x55effd935ad1]
supertux2(_ZN13ScreenManager14process_eventsEv+0xa5)[0x55effd9f2f65]
supertux2(_ZN13ScreenManager9loop_iterEv+0x18b)[0x55effd9f39eb]
supertux2(_ZN13ScreenManager3runEv+0x30)[0x55effd9f3c60]
supertux2(_ZN4Main11launch_gameERK20CommandLineArguments+0xae8)[0x55effd9b3978]
supertux2(_ZN4Main3runEiPPc+0x50b)[0x55effd9b4b4b]
supertux2(main+0x4e)[0x55effd8e623e]
/lib/x86_64-linux-gnu/libc.so.6(+0x29d90)[0x7f2415e86d90]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80)[0x7f2415e86e40]
supertux2(_start+0x2e)[0x55effd91961e]
Segmentation fault (core dumped)
```

**Comment by Frostwithasideofsalt** [comment]
image
crashes whenever i tap an object in the list of objects

**Comment by Frostwithasideofsalt** [comment]
there also seems to be duplicates of objects in the list

**Review by Rusty-Box — APPROVED** [review]

File: src/supertux/object_factory.hpp [hunk 1]
```diff
@@ -93,6 +98,12 @@ class ObjectFactory
                   return std::make_unique<C>(reader);
                 }, obj_params);
   }
+
+  template<class C>
+  void add_factory_compat(const char* class_name, uint8_t obj_params = 0, const std::string& display_name = "")
+  {
+    add_factory<C>(class_name, obj_params + RegisteredObjectParam::OBJ_PARAM_NON_DISPENSABLE, display_name);
```
Comment (mrkubax10) [inline src/supertux/object_factory.hpp:105]: You should use logical OR operator (`|`) instead of addition here.

**Review by mrkubax10 — APPROVED** [review]

File: src/supertux/game_object_factory.cpp [hunk 2]
```diff
@@ -160,151 +160,151 @@ GameObjectFactory::init_factories()
...
-  add_factory<RubLight>("rublight");
-  add_factory<ScriptedObject>("scriptedobject", RegisteredObjectParam::OBJ_PARAM_NON_DISPENSABLE);
-  add_factory<Shard>("shard");
+  add_factory<ParticleZone>("particle-zone");
+  add_factory<Platform>("platform");
+  add_factory<PneumaticPlatform>("pneumatic-platform");
+  add_factory<PowerUp>("powerup", OBJ_PARAM_DISPENSABLE);
+  add_factory<PushButton>("pushbutton", OBJ_PARAM_DISPENSABLE);
+  add_factory<RainParticleSystem>("particles-rain");
+  add_factory<Rock>("rock", OBJ_PARAM_PORTABLE & OBJ_PARAM_DISPENSABLE);
```
Comment (mrkubax10) [inline src/supertux/game_object_factory.cpp:270]: You should use OR there

File: src/supertux/game_object_factory.cpp [hunk 3]
```diff
@@ -160,151 +160,151 @@ GameObjectFactory::init_factories()
...
-  add_factory<TextArrayObject>("text-array", RegisteredObjectParam::OBJ_PARAM_NON_DISPENSABLE);
-  add_factory<Thunderstorm>("thunderstorm", RegisteredObjectParam::OBJ_PARAM_NON_DISPENSABLE);
+  add_factory<TextScroller>("textscroller");
+  add_factory<TextArrayObject>("text-array");
+  add_factory<Thunderstorm>("thunderstorm");
   add_factory<Torch>("torch");
-  add_factory<Trampoline>("trampoline", RegisteredObjectParam::OBJ_PARAM_PORTABLE);
-  add_factory<RustyTrampoline>("rustytrampoline", RegisteredObjectParam::OBJ_PARAM_PORTABLE);
+  add_factory<Trampoline>("trampoline", OBJ_PARAM_PORTABLE & OBJ_PARAM_DISPENSABLE);
+  add_factory<RustyTrampoline>("rustytrampoline", OBJ_PARAM_PORTABLE & OBJ_PARAM_DISPENSABLE);
```
Comment (mrkubax10) [inline src/supertux/game_object_factory.cpp:282]: Same as above

**Pull request merged**